Libraries/Opc.Ua.Configuration/IApplicationInstanceBuilder.cs

```diff
@@ -0,0 +1,167 @@
+        IApplicationInstanceBuilderServerSelected AsServer(
+            string[] baseAddresses,
+            string[] alternateBaseAddresses = null);
+    }
+
+    /// <summary>
+    /// The server configuration.
+    /// </summary>
+    public interface IApplicationInstanceBuilderClient
+    {
+        /// <summary>
+        /// Configure instance to be used for UA client.
+        /// </summary>
+        IApplicationInstanceBuilderClientSelected AsClient();
+    }
+
+    /// <summary>
+    /// 
+    /// </summary>
+    public interface IApplicationInstanceBuilderSecurity
+    {
+        /// <summary>
+        /// Add the security configuration.
+        /// </summary>
+        /// <remarks>
+        /// The pki root path default to the directory store
+        /// location as defined in <see cref="CertificateStoreIdentifier.DefaultPKIRoot"/>
+        /// A <see cref="CertificateStoreType"/> defaults to the corresponding default store location.
+        /// </remarks>
+        /// <param name="subjectName">Application certificate subject name as distinguished name. A DC=localhost entry is converted to the hostname. The common name CN= is mandatory.</param>
+        /// <param name="pkiRoot">The path to the pki root.</param>
+        /// <param name="appRoot">The path to the app cert store, if different than the pki root.</param>
+        /// <param name="rejectedRoot">The path to the rejected certificate store.</param>
+        IApplicationInstanceBuilderCreate AddSecurityConfiguration(
+            string subjectName,
+            string pkiRoot = null,
+            string appRoot = null,
+            string rejectedRoot = null
+            );
+    }
+
+    /// <summary>
+    /// 
+    /// </summary>
+    public interface IApplicationInstanceBuilderServerPolicies
+    {
+        /// <summary>
+        /// Add the unsecure security policy type none to server configuration.
+        /// </summary>
+        IApplicationInstanceBuilderServerSelected AddUnsecurePolicyNone();
+
+        /// <summary>
+        /// Add the sign security policies to the server configuration.
+        /// </summary>
+        IApplicationInstanceBuilderServerSelected AddSignPolicies(bool deprecated = false);
+
+        /// <summary>
+        /// Add the sign and encrypt security policies to the server configuration.
+        /// </summary>
+        IApplicationInstanceBuilderServerSelected AddSignAndEncryptPolicies(bool deprecated = false);
+
+        /// <summary>
+        /// Add user token policy to server config.
+        /// </summary>
+        /// <param name="userTokenType">The user token type to add.</param>
+        /// <param name="clean">replace policies before add.</param>
+        IApplicationInstanceBuilderServerSelected AddUserTokenPolicy(UserTokenType userTokenType, bool clean = false);
+    }
+
+    /// <summary>
+    /// Create and validate the application configuration.
+    /// </summary>
+    public interface IApplicationInstanceBuilderCreate
```

Comment:
regarding the naming, maybe IApplicationConfigurationBuilder is more accurate

Reply:
indeed...